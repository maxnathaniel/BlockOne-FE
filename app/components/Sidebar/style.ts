import styled from '../../styles/styled-components';

interface ISidebarContainerProps {
  position: string;
  height: string;
  width: string;
}

interface IMenuItemsProps {
  isSelected: boolean;
}

export const Logo = styled.div`
  min-height: 200px;
  vertical-align: center;
  text-align: center;
`;

export const MenuItemContainer = styled.ul`
  margin: 0;
  padding: 0;

  @media all and (max-width: 800px) {
    display: flex;
    flex: 1 1 auto;
  }
`;

export const MenuItem = styled.li<IMenuItemsProps>`
  display: block;
  list-style: none;
  line-height: 50px;
  text-align: center;
  margin: 0;
  padding: 5px;
  color: #E0E3ED;
  background-color: ${props => props.isSelected ? '#272F4D' : '#131930'};

  &: hover {
    cursor: pointer;
    background-color: #272F4D;
  }

  > a {
    display: block;
    line-height: 50px;
    padding: 5px;
    color: #66FCF1;
  }

  @media all and (max-width: 800px) {
    display: inline-block;
    flex: 1 1 auto;
    min-width: max-content;

    > a {
      line-height: 20px;
    }
  }
`;

export const SidebarContainer: any = styled.div<ISidebarContainerProps>`
  position: absolute;
  height: 100%;
  width: 15%;
  background-color: #131930;

  @media all and (max-width: 800px) {
    bottom: 0;
    width: 100%;
    display: flex;
    height: 40px;
    z-index: 999;
    overflow: scroll;
  }
`;
