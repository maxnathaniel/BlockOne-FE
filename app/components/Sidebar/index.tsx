import React, { useState, useCallback, useEffect, useReducer } from 'react';
import { Link } from 'react-router-dom';
import { MENU_ITEMS } from '../../constants';
import { MenuItem, MenuItemContainer, SidebarContainer } from './style';

const Sidebar: React.FC<{}> = () => {
  const URLs = {
    ['/']: true,
    ['/transactions']: false,
    ['/services']: false,
    ['/investment']: false,
    ['/loans']: false
  };

  const [urls, setUrls] = useState<any>(URLs);

  const onClick = useCallback((e: any) => {
    const menuItem = MENU_ITEMS[e.target.getAttribute('data-val')];
    const clickedItem = menuItem.key;
    // although this method of seriaizing and pasrsing the string creates
    // a deep copy object, note that it only works for simple object
    // for more complex object types, do use a library method
    const URLsCopy = JSON.parse(JSON.stringify(urls));

    for (const i in URLsCopy) {
      if (URLsCopy.hasOwnProperty(i)) {
        URLsCopy[i] = false;
        if (i === clickedItem) {
          URLsCopy[i] = true;
        }
      }
    }
    setUrls(URLsCopy);
  }, [urls]);

  return (
    <SidebarContainer>
      <MenuItemContainer>
        {
          Object.keys(MENU_ITEMS).map(item => (
            <MenuItem key={MENU_ITEMS[item].key} isSelected={urls[MENU_ITEMS[item].key]}>
              <Link data-val={item} to={`${MENU_ITEMS[item].key}`} onClick={onClick}>{MENU_ITEMS[item].name}</Link>
            </MenuItem>
          ))
        }
      </MenuItemContainer>
    </SidebarContainer>
  );
};

export default Sidebar;