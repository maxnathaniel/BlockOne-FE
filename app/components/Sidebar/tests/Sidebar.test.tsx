import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { MemoryRouter } from 'react-router-dom';
import Sidebar from '../index';

describe('<Sidebar />', () => {
  it('should render the menu items', () => {
    const { container, getByText }
      = render(<MemoryRouter><Sidebar /></MemoryRouter>);
    expect(getByText('Accounts')).toBeInTheDocument();
    expect(getByText('Transactions')).toBeInTheDocument();
    expect(getByText('Services')).toBeInTheDocument();
    expect(getByText('Investment')).toBeInTheDocument();
    expect(getByText('Loans')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});