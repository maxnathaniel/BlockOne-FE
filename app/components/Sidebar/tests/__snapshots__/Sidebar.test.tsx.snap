// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> should render the menu items 1`] = `
<div>
  <div
    class="sc-AykKF CoKPR"
  >
    <ul
      class="sc-AykKD hrORuh"
    >
      <li
        class="sc-AykKE hUuNdQ"
      >
        <a
          data-val="ACCOUNTS"
          href="/"
        >
          Accounts
        </a>
      </li>
      <li
        class="sc-AykKE tANiE"
      >
        <a
          data-val="TRANSACTIONS"
          href="/transactions"
        >
          Transactions
        </a>
      </li>
      <li
        class="sc-AykKE tANiE"
      >
        <a
          data-val="SERVICES"
          href="/services"
        >
          Services
        </a>
      </li>
      <li
        class="sc-AykKE tANiE"
      >
        <a
          data-val="INVESTMENT"
          href="/investment"
        >
          Investment
        </a>
      </li>
      <li
        class="sc-AykKE tANiE"
      >
        <a
          data-val="LOANS"
          href="/loans"
        >
          Loans
        </a>
      </li>
    </ul>
  </div>
</div>
`;
